File: clients/go-sdk/client/clients.go
package client

import (
	"context"
	"fmt"
)

// ─── Request DTOs ────────────────────────────────────────────────────

// CreateClientRequest — POST /api/clients.
type CreateClientRequest struct {
	Identifier string `json:"identifier"`
	Name       string `json:"name"`
}

// UpdateClientRequest — PUT /api/clients/{id}.
type UpdateClientRequest struct {
	Name *string `json:"name,omitempty"`
}

// StatusChangeRequest — body for suspend / deactivate.
type StatusChangeRequest struct {
	Reason string `json:"reason"`
}

// AddNoteRequest — body for POST /api/clients/{id}/notes.
type AddNoteRequest struct {
	Category string `json:"category"`
	Text     string `json:"text"`
}

// UpdateClientApplicationsRequest — bulk enable/disable.
type UpdateClientApplicationsRequest struct {
	EnabledApplicationIDs []string `json:"enabledApplicationIds"`
}

// ─── Response DTOs ───────────────────────────────────────────────────

// ClientResponse is the platform's client (tenant) aggregate.
type ClientResponse struct {
	ID              string `json:"id"`
	Name            string `json:"name"`
	Identifier      string `json:"identifier"`
	Status          string `json:"status"`
	StatusReason    string `json:"statusReason,omitempty"`
	StatusChangedAt string `json:"statusChangedAt,omitempty"`
	CreatedAt       string `json:"createdAt"`
	UpdatedAt       string `json:"updatedAt"`
}

// ClientListResponse — GET /api/clients.
type ClientListResponse struct {
	Clients []ClientResponse `json:"clients"`
	Total   uint64           `json:"total,omitempty"`
}

// StatusChangeResponse — { message } envelope from status-change endpoints.
type StatusChangeResponse struct {
	Message string `json:"message"`
}

// AddNoteResponse — { message } envelope from add-note endpoint.
type AddNoteResponse struct {
	Message string `json:"message"`
}

// ClientApplicationResponse — one entry in GET /api/clients/{id}/applications.
type ClientApplicationResponse struct {
	ID               string `json:"id"`
	Code             string `json:"code"`
	Name             string `json:"name"`
	Description      string `json:"description,omitempty"`
	IconURL          string `json:"iconUrl,omitempty"`
	Active           bool   `json:"active,omitempty"`
	EnabledForClient bool   `json:"enabledForClient,omitempty"`
}

// ClientApplicationsResponse — GET /api/clients/{id}/applications.
type ClientApplicationsResponse struct {
	Applications []ClientApplicationResponse `json:"applications"`
	Total        uint64                      `json:"total,omitempty"`
}

// ─── Resource ────────────────────────────────────────────────────────

// ClientsResource — /api/clients/*.
type ClientsResource struct {
	c *FlowCatalystClient
}

// Create — POST /api/clients.
func (r *ClientsResource) Create(ctx context.Context, req *CreateClientRequest) (*CreatedResponse, error) {
	var out CreatedResponse
	if err := r.c.Post(ctx, "/api/clients", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// List — GET /api/clients with optional status + pagination filters.
func (r *ClientsResource) List(ctx context.Context, status string, page, pageSize *uint32) (*ClientListResponse, error) {
	q := NewQuery().
		String("status", status).
		Uint32("page", page).
		Uint32("pageSize", pageSize).
		Encode()
	var out ClientListResponse
	if err := r.c.Get(ctx, "/api/clients"+q, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Get — GET /api/clients/{id}.
func (r *ClientsResource) Get(ctx context.Context, id string) (*ClientResponse, error) {
	var out ClientResponse
	if err := r.c.Get(ctx, "/api/clients/"+id, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// GetByIdentifier — GET /api/clients/by-identifier/{identifier}.
func (r *ClientsResource) GetByIdentifier(ctx context.Context, identifier string) (*ClientResponse, error) {
	var out ClientResponse
	if err := r.c.Get(ctx, "/api/clients/by-identifier/"+identifier, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Search — GET /api/clients/search?q=...
func (r *ClientsResource) Search(ctx context.Context, query string) (*ClientListResponse, error) {
	q := NewQuery().String("q", query).Encode()
	var out ClientListResponse
	if err := r.c.Get(ctx, "/api/clients/search"+q, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Update — PUT /api/clients/{id}.
func (r *ClientsResource) Update(ctx context.Context, id string, req *UpdateClientRequest) (*ClientResponse, error) {
	var out ClientResponse
	if err := r.c.Put(ctx, "/api/clients/"+id, req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Delete — DELETE /api/clients/{id}.
func (r *ClientsResource) Delete(ctx context.Context, id string) error {
	return r.c.Delete(ctx, "/api/clients/"+id, nil)
}

// Activate — POST /api/clients/{id}/activate.
func (r *ClientsResource) Activate(ctx context.Context, id string) (*StatusChangeResponse, error) {
	var out StatusChangeResponse
	if err := r.c.Post(ctx, "/api/clients/"+id+"/activate", nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Suspend — POST /api/clients/{id}/suspend.
func (r *ClientsResource) Suspend(ctx context.Context, id string, req *StatusChangeRequest) (*StatusChangeResponse, error) {
	var out StatusChangeResponse
	if err := r.c.Post(ctx, "/api/clients/"+id+"/suspend", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// Deactivate — POST /api/clients/{id}/deactivate.
func (r *ClientsResource) Deactivate(ctx context.Context, id string, req *StatusChangeRequest) (*StatusChangeResponse, error) {
	var out StatusChangeResponse
	if err := r.c.Post(ctx, "/api/clients/"+id+"/deactivate", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// AddNote — POST /api/clients/{id}/notes.
func (r *ClientsResource) AddNote(ctx context.Context, id string, req *AddNoteRequest) (*AddNoteResponse, error) {
	var out AddNoteResponse
	if err := r.c.Post(ctx, "/api/clients/"+id+"/notes", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// ListApplications — GET /api/clients/{id}/applications (with enabled status).
func (r *ClientsResource) ListApplications(ctx context.Context, clientID string) (*ClientApplicationsResponse, error) {
	var out ClientApplicationsResponse
	if err := r.c.Get(ctx, "/api/clients/"+clientID+"/applications", &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// EnableApplication — POST /api/clients/{clientId}/applications/{applicationId}/enable.
func (r *ClientsResource) EnableApplication(ctx context.Context, clientID, applicationID string) (*SuccessResponse, error) {
	var out SuccessResponse
	if err := r.c.Post(ctx, fmt.Sprintf("/api/clients/%s/applications/%s/enable", clientID, applicationID), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// DisableApplication — POST /api/clients/{clientId}/applications/{applicationId}/disable.
func (r *ClientsResource) DisableApplication(ctx context.Context, clientID, applicationID string) (*SuccessResponse, error) {
	var out SuccessResponse
	if err := r.c.Post(ctx, fmt.Sprintf("/api/clients/%s/applications/%s/disable", clientID, applicationID), nil, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

// UpdateApplications — PUT /api/clients/{id}/applications (bulk enable list).
func (r *ClientsResource) UpdateApplications(ctx context.Context, clientID string, req *UpdateClientApplicationsRequest) (*SuccessResponse, error) {
	var out SuccessResponse
	if err := r.c.Put(ctx, "/api/clients/"+clientID+"/applications", req, &out); err != nil {
		return nil, err
	}
	return &out, nil
}

File: crates/fc-sdk/src/client/clients.rs
//! Client (tenant) management operations.

use super::applications::{CreatedResponse, SuccessResponse};
use super::{ClientError, FlowCatalystClient};
use serde::{Deserialize, Serialize};

/// Request to create a client (tenant).
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct CreateClientRequest {
    /// Unique identifier for the client (e.g., slug or domain)
    pub identifier: String,
    /// Human-readable name
    pub name: String,
}

/// Request to update a client.
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct UpdateClientRequest {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
}

/// Request for status change operations (suspend, deactivate).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusChangeRequest {
    /// Reason for the status change
    pub reason: String,
}

/// Request to add a note to a client.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddNoteRequest {
    /// Note category
    pub category: String,
    /// Note text
    pub text: String,
}

/// Request to update client applications (bulk enable/disable).
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct UpdateClientApplicationsRequest {
    /// Application IDs to enable for this client
    pub enabled_application_ids: Vec<String>,
}

/// Client response from the platform API.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientResponse {
    pub id: String,
    pub name: String,
    pub identifier: String,
    pub status: String,
    #[serde(default)]
    pub status_reason: Option<String>,
    #[serde(default)]
    pub status_changed_at: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

/// Client list response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientListResponse {
    pub clients: Vec<ClientResponse>,
    #[serde(default)]
    pub total: Option<u64>,
}

/// Status change response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct StatusChangeResponse {
    pub message: String,
}

/// Add note response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AddNoteResponse {
    pub message: String,
}

/// Client application status response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientApplicationResponse {
    pub id: String,
    pub code: String,
    pub name: String,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub icon_url: Option<String>,
    #[serde(default)]
    pub active: bool,
    #[serde(default)]
    pub enabled_for_client: bool,
}

/// Client applications list response.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ClientApplicationsResponse {
    pub applications: Vec<ClientApplicationResponse>,
    #[serde(default)]
    pub total: Option<u64>,
}

/// Clients resource accessor — created via [`FlowCatalystClient::clients`].
pub struct Clients<'a> {
    pub(crate) client: &'a FlowCatalystClient,
}

impl Clients<'_> {
    /// Create a new client (tenant).
    pub async fn create(
        &self,
        req: &CreateClientRequest,
    ) -> Result<CreatedResponse, ClientError> {
        self.client.post("/api/clients", req).await
    }

    /// List clients with optional pagination and status filter.
    pub async fn list(
        &self,
        status: Option<&str>,
        page: Option<u32>,
        page_size: Option<u32>,
    ) -> Result<ClientListResponse, ClientError> {
        let mut params = Vec::new();
        if let Some(s) = status {
            params.push(format!("status={}", s));
        }
        if let Some(p) = page {
            params.push(format!("page={}", p));
        }
        if let Some(ps) = page_size {
            params.push(format!("pageSize={}", ps));
        }

        let query = if params.is_empty() {
            String::new()
        } else {
            format!("?{}", params.join("&"))
        };

        self.client.get(&format!("/api/clients{}", query)).await
    }

    /// Get a client by ID.
    pub async fn get(&self, id: &str) -> Result<ClientResponse, ClientError> {
        self.client.get(&format!("/api/clients/{}", id)).await
    }

    /// Get a client by identifier (slug/domain).
    pub async fn get_by_identifier(
        &self,
        identifier: &str,
    ) -> Result<ClientResponse, ClientError> {
        self.client
            .get(&format!("/api/clients/by-identifier/{}", identifier))
            .await
    }

    /// Search clients by name or identifier.
    pub async fn search(&self, query: &str) -> Result<ClientListResponse, ClientError> {
        self.client
            .get(&format!("/api/clients/search?q={}", query))
            .await
    }

    /// Update a client.
    pub async fn update(
        &self,
        id: &str,
        req: &UpdateClientRequest,
    ) -> Result<ClientResponse, ClientError> {
        self.client.put(&format!("/api/clients/{}", id), req).await
    }

    /// Delete (deactivate) a client.
    pub async fn delete(&self, id: &str) -> Result<(), ClientError> {
        self.client.delete_req(&format!("/api/clients/{}", id)).await
    }

    /// Activate a client.
    pub async fn activate(&self, id: &str) -> Result<StatusChangeResponse, ClientError> {
        self.client
            .post_action(&format!("/api/clients/{}/activate", id))
            .await
    }

    /// Suspend a client.
    pub async fn suspend(
        &self,
        id: &str,
        req: &StatusChangeRequest,
    ) -> Result<StatusChangeResponse, ClientError> {
        self.client
            .post(&format!("/api/clients/{}/suspend", id), req)
            .await
    }

    /// Deactivate a client.
    pub async fn deactivate(
        &self,
        id: &str,
        req: &StatusChangeRequest,
    ) -> Result<StatusChangeResponse, ClientError> {
        self.client
            .post(&format!("/api/clients/{}/deactivate", id), req)
            .await
    }

    /// Add a note to a client.
    pub async fn add_note(
        &self,
        id: &str,
        req: &AddNoteRequest,
    ) -> Result<AddNoteResponse, ClientError> {
        self.client
            .post(&format!("/api/clients/{}/notes", id), req)
            .await
    }

    /// List applications for a client (with enabled status).
    pub async fn list_applications(
        &self,
        client_id: &str,
    ) -> Result<ClientApplicationsResponse, ClientError> {
        self.client
            .get(&format!("/api/clients/{}/applications", client_id))
            .await
    }

    /// Enable an application for a client.
    pub async fn enable_application(
        &self,
        client_id: &str,
        application_id: &str,
    ) -> Result<SuccessResponse, ClientError> {
        self.client
            .post_action(&format!(
                "/api/clients/{}/applications/{}/enable",
                client_id, application_id
            ))
            .await
    }

    /// Disable an application for a client.
    pub async fn disable_application(
        &self,
        client_id: &str,
        application_id: &str,
    ) -> Result<SuccessResponse, ClientError> {
        self.client
            .post_action(&format!(
                "/api/clients/{}/applications/{}/disable",
                client_id, application_id
            ))
            .await
    }

    /// Bulk update which applications are enabled for a client.
    pub async fn update_applications(
        &self,
        client_id: &str,
        req: &UpdateClientApplicationsRequest,
    ) -> Result<SuccessResponse, ClientError> {
        self.client
            .put(&format!("/api/clients/{}/applications", client_id), req)
            .await
    }
}
